src/arena.c

```diff
@@ -37,6 +37,207 @@ static void	arena_bin_lower_slab(tsdn_t *tsdn, arena_t *arena,
+	    NBINS].nrequests, nrequests);
+	arena_stats_unlock(tsdn, arena_stats);
+}
+
+void
+arena_basic_stats_merge(tsdn_t *tsdn, arena_t *arena, unsigned *nthreads,
+    const char **dss, ssize_t *decay_time, size_t *nactive, size_t *ndirty) {
+	malloc_mutex_lock(tsdn, &arena->lock);
+	*nthreads += arena_nthreads_get(arena, false);
+	*dss = dss_prec_names[arena_dss_prec_get(arena)];
+	*decay_time = arena->decay.time;
+	*nactive += atomic_read_zu(&arena->nactive);
+	*ndirty += extents_npages_get(&arena->extents_cached);
+	malloc_mutex_unlock(tsdn, &arena->lock);
+}
+
+void
+arena_stats_merge(tsdn_t *tsdn, arena_t *arena, unsigned *nthreads,
+    const char **dss, ssize_t *decay_time, size_t *nactive, size_t *ndirty,
+    arena_stats_t *astats, malloc_bin_stats_t *bstats,
+    malloc_large_stats_t *lstats) {
+	size_t base_allocated, base_resident, base_mapped;
+	unsigned i;
+
+	cassert(config_stats);
+
+	arena_basic_stats_merge(tsdn, arena, nthreads, dss, decay_time,
+	    nactive, ndirty);
+
+	base_stats_get(tsdn, arena->base, &base_allocated, &base_resident,
+	    &base_mapped);
+
+	arena_stats_lock(tsdn, &arena->stats);
+
+	astats->mapped += base_mapped + arena_stats_read_zu(tsdn, &arena->stats,
+	    &arena->stats.mapped);
+	astats->retained += (extents_npages_get(&arena->extents_retained) <<
+	    LG_PAGE);
+	astats->npurge += arena_stats_read_u64(tsdn, &arena->stats,
+	    &arena->stats.npurge);
+	astats->nmadvise += arena_stats_read_u64(tsdn, &arena->stats,
+	    &arena->stats.nmadvise);
+	astats->purged += arena_stats_read_u64(tsdn, &arena->stats,
+	    &arena->stats.purged);
+	astats->base += base_allocated;
+	astats->internal += arena_internal_get(arena);
+	astats->resident += base_resident + (((atomic_read_zu(&arena->nactive) +
+	    extents_npages_get(&arena->extents_cached)) << LG_PAGE));
+	astats->allocated_large += arena_stats_read_zu(tsdn, &arena->stats,
+	    &arena->stats.allocated_large);
+	astats->nmalloc_large += arena_stats_read_u64(tsdn, &arena->stats,
+	    &arena->stats.nmalloc_large);
+	astats->ndalloc_large += arena_stats_read_u64(tsdn, &arena->stats,
+	    &arena->stats.ndalloc_large);
+	astats->nrequests_large += arena_stats_read_u64(tsdn, &arena->stats,
+	    &arena->stats.nrequests_large);
+
+	for (i = 0; i < NSIZES - NBINS; i++) {
+		lstats[i].nmalloc += arena_stats_read_u64(tsdn, &arena->stats,
+		    &arena->stats.lstats[i].nmalloc);
+		lstats[i].ndalloc += arena_stats_read_u64(tsdn, &arena->stats,
+		    &arena->stats.lstats[i].ndalloc);
+		lstats[i].nrequests += arena_stats_read_u64(tsdn, &arena->stats,
+		    &arena->stats.lstats[i].nrequests);
+		lstats[i].curlextents += arena_stats_read_zu(tsdn,
+		    &arena->stats, &arena->stats.lstats[i].curlextents);
+	}
+
+	arena_stats_unlock(tsdn, &arena->stats);
+
+	if (config_tcache) {
+		tcache_bin_t *tbin;
+		tcache_t *tcache;
+
+		/* tcache_bytes counts currently cached bytes. */
+		astats->tcache_bytes = 0;
+		ql_foreach(tcache, &arena->tcache_ql, link) {
```

Comment:
Yeah that's true. I was looking at them separately and forgot to discard that comment.